server/src/internalClusterTest/java/org/elasticsearch/gateway/GatewayServiceIT.java

```diff
@@ -0,0 +1,154 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.gateway;
+
+import org.elasticsearch.cluster.routing.ShardRouting;
+import org.elasticsearch.cluster.routing.allocation.AllocateUnassignedDecision;
+import org.elasticsearch.cluster.routing.allocation.ExistingShardsAllocator;
+import org.elasticsearch.cluster.routing.allocation.FailedShard;
+import org.elasticsearch.cluster.routing.allocation.RoutingAllocation;
+import org.elasticsearch.cluster.routing.allocation.decider.AllocationDecider;
+import org.elasticsearch.common.settings.ClusterSettings;
+import org.elasticsearch.common.settings.Setting;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.util.CollectionUtils;
+import org.elasticsearch.plugins.ClusterPlugin;
+import org.elasticsearch.plugins.Plugin;
+import org.elasticsearch.test.ESIntegTestCase;
+import org.elasticsearch.test.InternalTestCluster;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import static org.elasticsearch.test.hamcrest.ElasticsearchAssertions.assertAcked;
+
+@ESIntegTestCase.ClusterScope(scope = ESIntegTestCase.Scope.TEST)
+public class GatewayServiceIT extends ESIntegTestCase {
+
+    public static final Setting<Boolean> TEST_SETTING = Setting.boolSetting(
+        "gateway.test.setting",
+        false,
+        Setting.Property.NodeScope,
+        Setting.Property.Dynamic
+    );
+
+    public static final String ALLOCATOR_NAME = "test-shards-allocator";
+
+    public static class TestPlugin extends Plugin implements ClusterPlugin {
+
+        private final AtomicBoolean settingApplied = new AtomicBoolean();
+
+        @Override
+        public List<Setting<?>> getSettings() {
+            return List.of(TEST_SETTING);
+        }
+
+        @Override
+        public Collection<AllocationDecider> createAllocationDeciders(Settings settings, ClusterSettings clusterSettings) {
+            return List.of(new TestAllocationDecider(settings, clusterSettings, settingApplied));
+        }
+
+        @Override
+        public Map<String, ExistingShardsAllocator> getExistingShardsAllocators() {
+            return Map.of(ALLOCATOR_NAME, new ExistingShardsAllocator() {
+                @Override
+                public void beforeAllocation(RoutingAllocation allocation) {
+                    if (settingApplied.get() == false) {
+                        fail("boom");
+                    }
+                    assertTrue(settingApplied.get());
```

Comment:
I believe if statement and following assertion are doing exactly the same check.
I would suggest to only keep one and add a message describing the failure

Reply:
Ah whoops that was just so I had somewhere to hang a breakpoint. Fixed in 0ae1e299a32.